kernel/time/src/lib.rs

```diff
@@ -0,0 +1,316 @@
+            // fixed. This will probably be fixed if we separate interrupts from clocks in a
+            // future PR.
+            T::init()?;
+        }
+
+        EARLY_SLEEP_FUNCTION.store(T::sleep as usize, Ordering::SeqCst);
+        EARLY_SLEEPER_FREQUENCY.store(frequency, Ordering::SeqCst);
+
+        Ok(true)
+    } else {
+        Ok(false)
+    }
+}
+
+/// Wait for the given `duration`.
+///
+/// This function can be used even when interrupts are disabled.
+///
+/// # Panics
+///
+/// This function will panic if called prior to [`register_early_sleeper`].
+pub fn early_sleep(duration: Duration) {
+    let addr = EARLY_SLEEP_FUNCTION.load(Ordering::SeqCst);
+    if addr == 0 {
+        panic!("early sleep function not set");
+    } else {
+        let f: fn(Duration) = unsafe { core::mem::transmute(addr) };
+        f(duration)
+    }
+}
+
+/// Register a clock source.
+///
+/// The provided source will overwrite the previous source only if `frequency`
+/// is larger than that of the previous source.
+///
+/// Returns whether the previous source was overwritten.
+///
+/// # Errors
+///
+/// Returns an error if the source doesn't exist or if
+/// [`init`](ClockSource::init) returns an error.
+pub fn register_clock_source<T>(frequency: u64) -> Result<bool, &'static str>
+where
+    T: ClockSource,
+{
+    let old_frequency = T::ClockType::frequency_atomic().load(Ordering::SeqCst);
+    if frequency > old_frequency {
+        T::init()?;
+
+        let now_addr = T::ClockType::now_addr();
+        now_addr.store(T::now as usize, Ordering::SeqCst);
+
+        T::ClockType::store_instant_to_duration_func(T::unit_to_duration as usize);
+        T::ClockType::store_duration_to_instant_func(T::duration_to_unit as usize);
+
+        T::ClockType::frequency_atomic().store(frequency, Ordering::SeqCst);
+
+        Ok(true)
+    } else {
+        Ok(false)
+    }
+}
+
+/// Returns the current time.
+///
+/// Monotonic clocks return an [`Instant`] whereas realtime clocks return a
+/// [`Duration`] signifying the time since 12:00am January 1st 1970 (i.e. Unix
+/// time).
+///
+/// # Panics
+///
+/// This function will panic if called prior to registering a clock using
+/// [`register_clock_source`]. [`register_clock_source`] must return [`Ok`] and
+/// the [`ClockType`] of the registered clock must be the same as `T`.
+pub fn now<T>() -> T::Unit
+where
+    T: ClockType,
+{
+    let addr = T::now_addr().load(Ordering::SeqCst);
+    if addr == 0 {
+        panic!("time function not set");
+    } else {
+        let f: fn() -> T::Unit = unsafe { core::mem::transmute(addr) };
+        f()
+    }
+}
+
+/// A clock source.
+pub trait ClockSource {
+    /// The type of clock (either [`Monotonic`] or [`Realtime`]).
+    type ClockType: ClockType;
+
+    /// Whether the clock source exists on the system.
+    fn exists() -> bool;
+
+    /// Initialise the clock source.
+    fn init() -> Result<(), &'static str>;
```

Comment:
same question as above -- I thought we agreed to invert the control flow here to a more typical init routine:
* device manager invokes init routine for a given clock device
* clock device's init routine sets itself up, reading whatever config values from hardware it needs
* clock device's init routine registers itself as a clock, sets itself as a sleeper, etc

I don't think there's any reason for `exists()` to exist (heh), and `init()` also shouldn't be part of a `ClockSource` trait imo.